src/Accelerators/NNPA/Transform/ZLow/ZLowFinalPassRewrite.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------- ZLowRewrite.cpp - ZLow Rewrite Patterns ---------------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This pass implements optimizations for ZLow operations.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Interfaces/ViewLikeInterface.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "src/Accelerators/NNPA/Dialect/ZLow/ZLowOps.hpp"
+#include "src/Accelerators/NNPA/Pass/NNPAPasses.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+namespace zlow {
+
+/// This pattern rewrites
+/// ```mlir
+///   zlow.lstm(%input, %h0, %c0, %input_weights, %input_bias,
+///             %hidden_weights, %hidden_bias, %work_area, %shape,
+///             %hn_output, %cf_output,
+///             %direction, %return_all_steps, %prev_layer)
+///   ...
+///   zlow.lstm(%hn_output, %h0_, %c0_, %input_weights_, %input_bias_,
+///             %hidden_weights_, %hidden_bias_, %work_area_, %shape_,
+///             %hn_output_, %cf_output_,
+///             %direction_, %return_all_steps_, "none")
+/// ===>
+///   zlow.lstm(%input, %h0, %c0, %input_weights, %input_bias,
+///             %hidden_weights, %hidden_bias, %work_area, %shape,
+///             %hn_output, %cf_output,
+///             %direction, %return_all_steps, %prev_layer)
+///   ...
+///   zlow.lstm(%hn_output, %h0_, %c0_, %input_weights_, %input_bias_,
+///             %hidden_weights_, %hidden_bias_, %work_area_, %shape_,
+///             %hn_output_, %cf_output_,
+///             %direction_, %return_all_steps_,
+///             (%direction == "bidirectional) ? "bidir" : "uni") #was "none"
+///
+/// ```
+/// by changing the prev_layer parameter from "none" to "bidir" or "uni"
+/// according to the operation of the input argument.
+///   "none" if input tensor is not from a previous RNN layer
+///   "uni" if input tensor is uni-directional output from a previous RNN layer
+///   "bidir" if input tensor is bi-directional output from a previous RNN layer
+//
+class SetPrevLayerInLSTMOpPattern : public OpRewritePattern<ZLowLSTMOp> {
+public:
+  using OpRewritePattern<ZLowLSTMOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ZLowLSTMOp lstmOp, PatternRewriter &rewriter) const override {
+    Value lstmInput = lstmOp.input();
+    StringRef prevLayer = lstmOp.prev_layer();
+    if (strcmp(prevLayer.data(), "not_set")) // if prev_layer is set already
+      return failure();
+
+    // Search for LSTM/GRU op that generates the input argument.
+    StringRef directionAttr = "";
+    for (Operation *user : lstmInput.getUsers()) {
+      const char *opName = user->getName().getStringRef().data();
+      if (!strcmp(opName, "zlow.lstm")) {
```

Comment:
String comparing against all of the users is compile time expensive. I think you could instead attempt to use the `isa` member function ( e.g.  `user->isa<ZLowLSTMOp>()` ) here and in other places that currently use `strcmp`

Reply:
The "strcmp" function is replaced with "isa<ZLowLSTMOp>(user)." Thank you.